src/mlpack/methods/reinforcement_learning/async_learning_impl.hpp

```diff
@@ -0,0 +1,147 @@
+/**
+ * @file async_learning_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is the implementation of AsyncLearning class,
+ * which is wrapper for various asynchronous learning algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+#define MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "queue"
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::AsyncLearning(
+    TrainingConfig config,
+    NetworkType network,
+    PolicyType policy,
+    UpdaterType updater,
+    EnvironmentType environment):
+    config(std::move(config)),
+    learningNetwork(std::move(network)),
+    policy(std::move(policy)),
+    updater(std::move(updater)),
+    environment(std::move(environment))
+{ /* Nothing to do here. */ };
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+template <typename Measure>
+void AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::Train(Measure& measure)
+{
+  /**
+   * OpenMP doesn't support shared class member variables.
+   * So we need to copy them to local variables.
+   */
+  NetworkType learningNetwork = std::move(this->learningNetwork);
+  if (learningNetwork.Parameters().is_empty())
+    learningNetwork.ResetParameters();
+  NetworkType targetNetwork = learningNetwork;
+  size_t totalSteps = 0;
+  PolicyType policy = this->policy;
+  bool stop = false;
+
+  // Set up worker pool, worker 0 will be deterministic for evaluation.
+  std::vector<WorkerType> workers;
+  for (size_t i = 0; i <= config.NumWorkers(); ++i)
+  {
+    workers.push_back(WorkerType(updater, environment, config, !i));
+    workers.back().Initialize(learningNetwork);
+  }
+  // Set up task queue corresponding to worker pool.
+  std::queue<size_t> tasks;
+  for (size_t i = 0; i <= config.NumWorkers(); ++i)
+    tasks.push(i);
+
+  // Synchronization for get/put task.
+  omp_lock_t tasksLock;
```

Comment:
There is one last thing we should talk about that I missed, if OpenMP isn't supported the build would fail, so do you think we should surround the code with ``#ifdef HAS_OPENMP`` or just specific parts as I did in #1077. Let me know what you think.